File: dotnet/cs/example/Samples/Advanced.cs
/*
 * File: Advanced.cs
 * Project: Test
 * Created Date: 23/05/2021
 * Author: Shun Suzuki
 * -----
 * Last Modified: 12/09/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2022-2023 Shun Suzuki. All rights reserved.
 * 
 */

using AUTD3Sharp;
using AUTD3Sharp.Gain;
using AUTD3Sharp.Modulation;

namespace Samples;

public class AdvancedTest
{
    private class UniformGain : Gain
    {
        public override Dictionary<int, Drive[]> Calc(Geometry geometry)
        {
            return Transform(geometry, (_, _) => new Drive { Phase = 0.0, Amp = 1.0 });
        }
    }

    private class Burst : Modulation
    {
        private readonly int _length;

        public Burst(int length, double sampleFreqDiv = 5120) : base(sampleFreqDiv)
        {
            _length = length;
        }

        public override double[] Calc()
        {
            var buf = new double[_length];
            buf[0] = 1;
            return buf;
        }
    }

    public static void Test(Controller autd)
    {
        var config = Silencer.Disable();
        autd.Send(config);

        var g = new UniformGain();
        var m = new Burst(4000);

        autd.Send((m, g));
    }
}

File: dotnet/fs/example/Samples/Advanced.fs
// File: Advanced.fs
// Project: Samples
// Created Date: 03/02/2023
// Author: Shun Suzuki
// -----
// Last Modified: 12/09/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
// 

namespace Samples

open AUTD3Sharp
open AUTD3Sharp.Gain
open AUTD3Sharp.Modulation

module AdvancedTest =
    type UniformGain () =
        inherit Gain()
        let Calc_ (dev: Device) (tr:Transducer) = 
            let mutable drive = new Drive();
            drive.Amp <- 1.0;
            drive
        override this.Calc (geometry: Geometry) = Gain.Transform(geometry, Calc_);
        
    type Burst (size: int) =
        inherit Modulation(5120)
        override this.Calc () = 
            let buf: float array = Array.zeroCreate size
            buf[0] <- 1.
            buf
        
    let Test (autd : Controller) = 
        (Silencer.Disable()) |> autd.Send |> ignore

        let m = new Burst(4000);
        let g = new UniformGain();

        (m, g) |> autd.Send |> ignore
